selenium/src/main/java/com/deque/html/axecore/selenium/AxeBuilder.java

```diff
@@ -686,32 +702,53 @@ private ArrayList<String> runPartialRecursive(
       String resResponse =
           (String)
               WebDriverInjectorExtensions.executeAsyncScript(
-                  webDriver, runPartialScript, context, options);
-      ArrayList<String> partialResults = new ArrayList<String>();
+                  webDriver, runPartialScript, context, options, frameStack);
       partialResults.add(resResponse);
       if (disableIframeTesting) {
         return partialResults;
       }
 
       for (FrameContext fc : contexts) {
-        Object frameContext = AxeReporter.serialize(fc.getFrameContext());
-        Object frameSelector = AxeReporter.serialize(fc.getFrameSelector());
-        Object frame =
-            WebDriverInjectorExtensions.executeScript(webDriver, shadowSelectScript, frameSelector);
-
-        if (frame instanceof String) {
-          webDriver.switchTo().frame((String) frame);
-        } else if (frame instanceof WebElement) {
-          webDriver.switchTo().frame((WebElement) frame);
-        } else if (frame instanceof Integer) {
-          webDriver.switchTo().frame((Integer) frame);
-        } else {
+        try {
+          Object frameContext = AxeReporter.serialize(fc.getFrameContext());
+          Object frameSelector = AxeReporter.serialize(fc.getFrameSelector());
+          frameStack.push(frameSelector);

```

Comment:
Should only push to the `frameStack` when we've successfully `switchTo` the frame. Then you don't have to pop it later